tensorflow/stream_executor/matmul_util.h

```diff
@@ -0,0 +1,235 @@
+                           dtype_cd_, allow_tf32_, device_id_);
+  }
+
+  bool trans_a_;
+  bool trans_b_;
+  bool adj_a_;
+  bool adj_b_;
+  uint64 m_;
+  uint64 n_;
+  uint64 k_;
+  uint64 batch_count_;
+  bool broadcast_a_;
+  bool broadcast_b_;
+  tensorflow::DataType dtype_ab_;
+  tensorflow::DataType dtype_cd_;
+  bool allow_tf32_;
+  int device_id_;
+  uint64 hash_code_;
+};
+
+static inline port::StatusOr<blas::ComputationType> GetBlasComputationType(
+    const tensorflow::DataType& dtype, bool allow_tf32) {
+  using blas::ComputationType;
+  static bool use_f32_for_f16_computation =
+      tensorflow::MatmulDoFP32ComputationFP16Input();
+  ComputationType f32_type =
+      allow_tf32 ? ComputationType::kTF32AsF32 : ComputationType::kF32;
+  switch (dtype) {
+    case tensorflow::DT_HALF:
+    case tensorflow::DT_BFLOAT16:
+      return use_f32_for_f16_computation ? f32_type : ComputationType::kF16;
+    case tensorflow::DT_FLOAT:
+      return f32_type;
+    case tensorflow::DT_DOUBLE:
+      return ComputationType::kF64;
+    case tensorflow::DT_COMPLEX64:
+      return f32_type;
+    case tensorflow::DT_COMPLEX128:
+      return ComputationType::kComplexF64;
+    default:
+      return port::InternalError("Unsupported dtype for Blas Plans.");
+  }
+}
+
+static inline port::StatusOr<blas::DataType> GetBlasDataType(tensorflow::DataType dtype){
+  switch (dtype) {
+    case tensorflow::DT_HALF:
+      return blas::ToDataType<Eigen::half>::value;
+    case tensorflow::DT_FLOAT:
+      return blas::ToDataType<float>::value;
+    case tensorflow::DT_DOUBLE:
+      return blas::ToDataType<double>::value;
+    case tensorflow::DT_COMPLEX64:
+      return blas::ToDataType<tensorflow::complex64>::value;
+    case tensorflow::DT_COMPLEX128:
+      return blas::ToDataType<tensorflow::complex128>::value;
+    default:
+      return port::InternalError("Unsupported dtype for Blas Plans.");
+  }
+}
+
+// Thread-safe map from matmul parameters to their corresponding plan and
+// algorithms.
+template <typename Parameters>
+class BlasLtMatmulPlanMap {
+ public:
+  const blas::PlanAndAlgorithms* Find(const Parameters& params) {
+    tensorflow::mutex_lock lock(mu_);
+    auto iter = params_plan_map_.find(params);
+    if (iter == params_plan_map_.end()) {
+      return nullptr;
+    }
+    return &iter->second;
+  }
+  const blas::PlanAndAlgorithms* Insert(const Parameters& params,
+                                        blas::PlanAndAlgorithms value) {
+    tensorflow::mutex_lock lock(mu_);
+    return &params_plan_map_.emplace(params, std::move(value)).first->second;
+  }
+
+ private:
+  struct Hasher {
+    std::size_t operator()(const Parameters& parameter) const {
+      return parameter.hash();
+    }
+  };
+
+  tensorflow::mutex mu_;
+  std::unordered_map<Parameters, blas::PlanAndAlgorithms, Hasher>
+      params_plan_map_ GUARDED_BY(mu_);
+};
+
+template <typename Parameters>
+struct BlasLtPlanMapSingleton {
```

Comment:
Do you plan to add more parameterizations? If it's only ever used with one template parameterization, maybe it does not need templates at all?